In [1]:
import numpy as np

from sklearn.metrics import mean_squared_error as mse

1. Подбор скорости обучения (alpha) и количеста итераций:

In [63]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])

y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

n = X.shape[1]
accuracy = 0.01
alpha_list = [0.1, 0.01, 0.001]

In [69]:
for alpha in alpha_list:
    W = np.array([1, 0.5])
    distance = 1
    i = 0
    
    y_pred = np.dot(W, X)
    err0 = mse(y, y_pred)
    
    while distance > accuracy and distance > 0:
        i+=1
        W -= (alpha * (1/n * 2 * np.dot(X,(y_pred - y))))
        y_pred = np.dot(W, X)
        err1 = mse(y, y_pred)
        distance = err0 - err1
        err0 = err1
        
    print(i,W,err1)

1 [12.02 34.5 ] 9346.1224
369 [46.30016541  4.08281105] 46.40446490754569
2588 [44.24571609  4.46218083] 50.73237380391926


Наилучшая скорость обучения 0.01, так как при ней за наименьшее количество итераций (369) достигается минимальная ошибка (46,4) с точностью до 0.01. 

*2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [18]:
for i in range(1000):
    y_pred = np.dot(W, X)
    err = mse(y, y_pred)
    '''for ii in range(W.shape[0]):
    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))'''
    W -= (alpha * (1/n * 2 * np.dot(X,(y_pred - y)))) #np.dot вместо np.sum
    if i % 100 == 0:
        print(i, W, err)

0 [13.23083108 10.18305876] 667.560337269993
100 [13.58283629 10.11968975] 654.7658933898567
200 [13.93112726 10.05658129] 642.234896574031
300 [14.27576142  9.99383363] 629.96187344066
400 [14.61679067  9.93151919] 617.9414864469112
500 [14.95426267  9.86968976] 606.1685205795002
600 [15.28822192  9.80838193] 594.6378750864831
700 [15.61871044  9.74762106] 583.3445579848469
800 [15.94576831  9.68742425] 572.2836821052715
900 [16.26943415  9.62780246] 561.4504619966559
